File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.java
class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        Map<Character, Integer> map = new HashMap<>();
        List<Integer> result = new ArrayList<>();
        if(p.length() > s.length()) return result;
        int count = p.length();

        for(int i=0; i< count; i++) {
            if(!map.containsKey(p.charAt(i))) {
                map.put(p.charAt(i), 1);
            } else {
                map.put(p.charAt(i), map.get(p.charAt(i)) + 1);
            }
        }
        int left = 0, right = 0;
        
        while(right < p.length()) {
            if(map.containsKey(s.charAt(right))) {
                map.put(s.charAt(right), map.get(s.charAt(right)) - 1);
                if(map.get(s.charAt(right)) >= 0) count--;
            }

            right++;
        }

        if(count == 0) result.add(left);

        while(right < s.length()) {

            if(map.containsKey(s.charAt(left))) {
                map.put(s.charAt(left), map.get(s.charAt(left)) + 1);
                if(map.get(s.charAt(left)) > 0) count++;
            }
            left++;

            if(map.containsKey(s.charAt(right))) {
                map.put(s.charAt(right), map.get(s.charAt(right)) - 1);
                if(map.get(s.charAt(right)) >= 0) count--;
            }
            if(count == 0) result.add(left);
            right++;
        }

        return result;
    }
}

File: 438-find-all-anagrams-in-a-string/find-all-anagrams-in-a-string.cs
public class Solution {
    public IList<int> FindAnagrams(string s, string p) {
        if(s.Length < p.Length) return new List<int>();
        List<int> result = new List<int>();
        Dictionary<char, int> charMap = new Dictionary<char, int>();
        int count = p.Length;
        for (int i = 0; i <= count - 1; i++ ) {
            if(charMap.ContainsKey(p[i])) {
                int c = charMap[p[i]];
                charMap[p[i]] = c+1;
                
            } else {
                charMap.Add(p[i], 1);
            }
        }
        int left =0;
        int right =0;
        
        while(right < p.Length) {
            if(charMap.ContainsKey(s[right])) {
                int c = charMap[s[right]];
                charMap[s[right]] = c-1;
                if(charMap[s[right]] >= 0) count--;
            }
            right++;
        }

        if(count == 0) 
            result.Add(left);

        while(right < s.Length) {
            if(charMap.ContainsKey(s[left])) {
                int c = charMap[s[left]];
                charMap[s[left]] = c+1;
                if(charMap[s[left]] > 0) count++;
            }

            left++;

            if(charMap.ContainsKey(s[right])) {
                int c = charMap[s[right]];
                charMap[s[right]] = c-1;
                if(charMap[s[right]] >= 0)
                    count--;
            }
            
            if(count == 0) {
                result.Add(left);
            }
                

            right++;
        }

        return result;
    }
}